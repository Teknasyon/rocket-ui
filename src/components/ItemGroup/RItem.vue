<script setup lang="ts">
import { computed, inject, withDefaults } from 'vue';

export interface ItemProps {
  /**
   * The value of the item
   * @default null
   * @type any
   * @example
   * <RItem :value="1" />
   */
  value: any;

  /**
   * Whether the item is disabled
   * @default false
   * @type boolean
   * @example
   * <RItem :disabled="true" />
   */
  disabled?: boolean;

  /**
   * The class to apply to the selected item
   * @default ''
   * @type string
   * @example
   * <RItem selectedClass="bg-blue-500 text-white" />
   */
  selectedClass?: string;
}

const RItemGroupSymbol = 'rocket-ui:r-item-group';

const props = withDefaults(defineProps<ItemProps>(), {
  disabled: false,
  selectedClass: '',
  value: null,
});

const isSelected = inject(`${RItemGroupSymbol}:isSelected`) as (
  id: number
) => boolean;

const classes = inject(`${RItemGroupSymbol}:selectedClass`, '') as string;
const selectedClass = computed(() => {
  return isSelected(props.value) && [classes, props.selectedClass];
});

const select = inject(`${RItemGroupSymbol}:select`) as (
  id: number,
  value: boolean
) => void;

const handleToggle = () => {
  if (props.disabled) return;
  select(props.value as number, !isSelected(props.value as number));
};

const handleSelect = () => {
  if (props.disabled) return;
  select(props.value as number, true);
};
</script>
<template>
  <slot
    :disabled="props.disabled"
    :isSelected="isSelected(props.value as number)"
    :select="handleSelect"
    :selectedClass="selectedClass"
    :toggle="handleToggle"
    :value="props.value"
  />
</template>
