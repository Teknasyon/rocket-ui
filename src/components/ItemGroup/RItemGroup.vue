<script setup lang="ts">
import { computed, provide, ref } from 'vue'

export interface ItemGroupProps {
  /**
   * The value of the selected items
   * @default []
   * @model
   * @type {number[] | string[]}
   * @example
   * <RItemGroup v-model="selectedItems" />
   */
  modelValue: number[] | string[]

  /**
   * The class to apply to the selected items
   * @default ''
   * @type {string}
   * @example
   * <RItemGroup selectedClass="bg-blue-500 text-white" />
   */
  selectedClass?: string | string[]

  /**
   * Whether the item group is disabled
   * @default false
   * @type {boolean}
   * @example
   * <RItemGroup :disabled="true" />
   */
  disabled?: boolean

  /**
   * Whether the item group is mandatory
   * @default false
   * @type {boolean}
   * @example
   * ```js
   * const selected = ref<number[]>([1]);
   * ```
   * ```html
   * <RItemGroup :mandatory="true" v-model="selected" />
   * ```
   */
  mandatory?: boolean

  /**
   * The maximum number of items that can be selected
   * @default 0
   * @type {number}
   * @example
   * <RItemGroup :max="5" />
   */
  max?: number

  /**
   * The tag to use for the item group
   * @default div
   * @type {string}
   * @example
   * <RItemGroup as="ul" />
   */
  as?: string

  /**
   * Whether the item group allows multiple selections
   * @default false
   * @type {boolean}
   * @example
   * <RItemGroup :multiple="true" />
   */
  multiple?: boolean
}

const props = withDefaults(defineProps<ItemGroupProps>(), {
  disabled: false,
  mandatory: false,
  max: 0,
  as: 'div',
  selectedClass: '',
  modelValue: () => [] as number[],
})

const emits = defineEmits(['update:modelValue'])

const RItemGroupSymbol = 'rocket-ui:r-item-group'

const tag = ref<string>(props.as ?? 'div')

const selectedItems = ref<number[] | string[]>(props.modelValue ?? [])

const selectedClass = computed(() => props.selectedClass)
provide(`${RItemGroupSymbol}:selectedClass`, selectedClass.value)

function isSelected(id: never) {
  return selectedItems.value.includes(id)
}
provide(`${RItemGroupSymbol}:isSelected`, isSelected)

function select(id: never, value: boolean) {
  if (props.disabled)
    return
  if (props.mandatory && selectedItems.value.length === 1 && value)
    return
  if (props.max && selectedItems.value.length === props.max && value)
    return
  if (props.multiple) {
    if (value)
      selectedItems.value.push(id)
    else
      selectedItems.value = selectedItems.value.filter(item => item !== id) as never[]
  }
  else {
    selectedItems.value = value ? [id] : []
  }

  emits('update:modelValue', selectedItems.value)
}
provide(`${RItemGroupSymbol}:select`, select)
</script>

<template>
  <component :is="tag">
    <slot :is-selected="isSelected" :select="select" :selected="selectedItems" />
  </component>
</template>
