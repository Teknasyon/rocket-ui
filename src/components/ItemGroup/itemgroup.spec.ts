import { describe, expect, it } from 'vitest'

import { mount } from '@vue/test-utils'

import { h } from 'vue'

import RItemGroup from './RItemGroup.vue'
import RItem from './RItem.vue'

function defaultSlot() {
  return [
    h(
      RItem,
      { value: 'foo' },
      {
        default: ({ toggle }: any) =>
          h('div', { id: 'item', onClick: toggle }, ['foo']),
      },
    ),
    h(
      RItem,
      { value: 'bar' },
      {
        default: ({ toggle }: any) =>
          h('div', { id: 'item', onClick: toggle }, ['bar']),
      },
    ),
  ]
}

describe('RItemGroup', () => {
  it('should render correctly', async () => {
    const wrapper = mount(RItemGroup, {
      slots: {
        default: defaultSlot,
      },
    })

    const items = wrapper.findAll('#item')

    await items[0].trigger('click')
    await items[1].trigger('click')

    const events = wrapper.emitted('update:modelValue')

    expect(events).toHaveLength(2)
    expect(events).toEqual([[['foo']], [['bar']]])
  })

  it('should not deselect value when using mandatory prop', async () => {
    const wrapper = mount(RItemGroup, {
      props: {
        mandatory: true,
        modelValue: 'foo',
      },
      slots: {
        default: defaultSlot,
      },
    })

    const items = wrapper.findAll('#item')

    await items[0].trigger('click')

    const events = wrapper.emitted('update:modelValue')

    expect(events).toHaveLength(1)
  })
})
