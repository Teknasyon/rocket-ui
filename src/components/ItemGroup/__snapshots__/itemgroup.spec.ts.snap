// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ItemGroup > should render correctly 1`] = `
"<div>
  <div>Item</div>
</div>"
`;

exports[`RItemGroup > should render correctly 1`] = `
"<div>
  <div>Item</div>
</div>"
`;
